source/mothership-design.md

```diff
@@ -0,0 +1,771 @@
+   binaries.
+
+ - To drain the compute fabric as quickly as possible: One performance-failure
+   condition of a POETS compute job is for the compute fabric (Network-On-Chip
+   for Tinsel) to become overloaded with packets sent from devices to their
+   supervisor, or to external processes. This situation is easily reached by
+   applications with devices that regularly report their results to their
+   supervisor device, and causes packets from normal devices to be delayed by
+   this traffic, significantly hampering execution time. This problem is abated
+   when the Mothership process drains packets destined for the supervisor as
+   quickly as possible, without significant compromise to its other features.
+
+ - To support multiple applications simultaneously running on its box: This
+   means applications may be loaded while one is already running, and means
+   that the Mothership process must support the loading of multiple supervisor
+   binaries.
+
+ - Support a debugging interface (UART, in the case of Tinsel): Problems with
+   applications can be diagnosed more easily with a functioning debugging
+   interface.
+
+ - Operate in conjunction with other Mothership processes (multibox)
+
+ - Respond reasonably quickly to instructions sent to it (over MPI): Note that,
+   while resolving traffic from the compute fabric is a high priority activity,
+   certain command-and-control (C&C) instructions are more important still,
+   including quitting, and killing an application. The Mothership process must
+   enact these commands in reasonable time, without significantly compromising
+   fabric-draining performance.
+
+ - Support swappable backends: In the future, to use the Orchestrator with
+   non-Tinsel backends. The Mothership process should support a common
+   interface to multiple fabrics. This feature is not immediately essential,
+   but should be designed around.
+
+NB: Terminology in this document:
+
+ - *Message*: An addressed item (`(P)Msg_p` depending on context) with some
+   payload that traverses the MPI network via the `CommonBase` interface. NB:
+   MPI messages sent/received in this way are asynchronous.
+
+ - *Packet*: An addressed item (`P_Pkt_t`) with some payload that traverses the
+   compute fabric.
+
+ - *Debug packet*: An addressed item sent over DebugLink (UART) connection in
+   the Tinsel backend (`P_Debug_Pkt_t`, see the Debugging section).
+
+ - *Thread*: POSIX thread running under the Mothership process (x86-land). NB:
+   Not a "thread" in the compute fabric.
+
+ - *Mothership*: Occasionally, I refer to the Mothership as a class (or
+   object/instance), and a process. This is legacy - documentation is written
+   to take people from the past to the present after all. I'll try to be
+   explicit wherever I use this term.
+
+Figure 1 shows the class structure diagram for the proposed Mothership design.
+
+![Mothership class structure diagram](images/mothership_data_structure.pdf)
```

Comment:
That's odd, they do for me:
![mothership-docx-figure](https://user-images.githubusercontent.com/12502071/81272181-118fc780-9045-11ea-9add-b35bf732b937.png)
Any idea how we might resolve this? Do you think a `doc` build instead of a `docx`  build would make a difference?

Reply:
OK, I have done some more digging into this, and it looks like Word is not happy with the images being in PDF format.

If I exported the PDF as a PNG and included that instead of the PDF, I can make it display in Word.

Libre office doesn't care either way.

So potential solution: change all image paths to *.png and add an extra step to the build to convert PDF to PNG.

Reply:
I'd rather not rasterise the PDFs if possible. Do you know of any vectorised formats that Word will accept?

Reply:
A quick google implies that Word does not play that nicely with vector graphics.

It may support .emf (may only be supported in Windows), .wmf and .eps, but it does not support PDF.

2016 may support SVG, but apparently it rasterises it and not everyone uses 2016...

Reply:
This has turned into much more of a pain than it should have done, however on a whim I tried something - Adobe's conversion of PDF to Docx. It produced a doc with readable images from the source with the PDF images.

Libreoffice and abi word supposedly have conversion optiosn that may work (they seem to hang in my WSL Ubuntu instance so I can't check)

'''
libreoffice --infilter=="writer_pdf_import" --headless \
--convert-to doc:"writer_pdf_Export" <filename>.pdf
'''

Is it worth looking at that rather than resorting to PNG?